spec.md

```diff
@@ -113,8 +116,9 @@ circuit :
   module Qux :
 ```
 
-A circuit that contains no public modules is trivially equivalent to a circuit that contains no modules.
-It is not enforced that a circuit has at least one public module, but it is expected that it does.
+A circuit that contains no public modules is trivially equivalent to an empty circuit.
+It is not enforced that a circuit has at least one public module.
+It is not enforced that the main module is public.
```

Comment:
These lines seem strange to me.

How can a circuit be empty if:

-  it requires a name, and
- the name matches a public module

It feels like we're trying to document a quirk of the implementation. Maybe it would help to comment on the intention of these rules.

Reply:
Maybe there's a better way of putting this...

The PR is trying to make `public` mean the same thing for all modules. If a module is public then you get a Verilog module with ports following the port lowering ABI. If a module is private, anything can happen to the module. It may be optimized away, it may get inlined, it may be duplicated a million times, etc.

The main module creates a discontinuity arising from making the main module implicitly `public`. If it is implicitly public, then either the `public` keyword has no effect or the public keyword is mandatory (or the public keyword cannot be applied to the main module, but it is implicitly `public` anyway).

The language in this section is trying to say that a circuit which has no public modules can be optimized to produce no Verilog.

WDYT?

For full clarity, this is an intermediate step to FIRRTL 5. Currently you can get the "bag o' dags" representation in FIRRTL 4 with:

```
circuit Container :
  public module Foo :
  public module Bar :
  module Container :
    inst foo of Foo
    invalidate foo

    inst bar of Bar
    invalidate Bar
```

or perhaps the following, though FIRRTL compilers don't really handle this well right now:

```
circuit Dummy :
  public module Foo :
  public module Bar :
  module Dummy :
```

In FIRRTL 5 we can drop the circuit entirely and just make this:

```
public module Foo:
public module Bar:
```